### ISSUE TITLE
Allow to interact with a calculation job while it is running.

### ISSUE THREAD

**Issue by yakutovicha** [aiidateam/aiida-core#3868]
Would it be possible to allow for developers to provide an alternative kill method that is specific to the code?

For instance, for cp2k one can do `touch EXIT` to gently stop the calculation. Similar thing is possible for QE.

Any thoughts on that?

**Comment by sphuber** [comment]
I take it the use-case for this would be to pre-emptively stop a molecular dynamics run? I think this is a feature that falls under a more general request: the ability to interact with a calculation job while it is running. @espenfl has also voiced this use-case where he wants to ideally be able to monitor a file during the calcjobs lifetime and indeed then, depending on the parsed output, perform an action such as interrupt the calculation. I think this is a common use case and certainly useful to have.

As I see it, the challenges are two-fold:

1. The technical implementation
2. The user interface

For the technical information, we more or less have most of the technology in place. Each calculation job goes through a couple of steps that are called "Transport Tasks". Currently they are:

- `Upload`
- `Submit`
- `Update`
- `Retrieve`

During the `Update` the daemon is periodically updating the status by polling the scheduler. All these transport tasks are coroutines such that when waiting they do not block the runner's thread and it can run multiple of these "concurrently". What we could try is to add another TT, let's say `Monitor`, that runs (if specified) as long as the job is in the `Update` task. With a to be configured interval it can run some function that performs an action over the transport. The tricky part here is just to make sure that this additional TT is properly restored upon reloading the process instance, after for example a daemon restart.

Secondly, I think the real difficult part is also going to be the user interface. Even though the concept is pretty generic, what needs to happen in specific cases can be very difficult. Typically there will be some action to "inspect" something and then perform an "action". Both the inspection and action will be highly case specific. For example, the inspection can be to retrieve a particular file and then parse it, but if that file is huge, one might not want to retrieve the file every interval. Then the parsing should happen remotely, so the inspection instead would be the execution of a command on the remote.

In summary, I think it is a super useful feature to have, but when thinking on how to implement this in a generic way, it can be quite challenging.

**Comment by espenfl** [comment]
Yes, this is rather important functionality to get in place. As soon as we have the scheduler parsing completed this would be the next on the list. E.g. to be able to kill, given some detected error (VASP very often just continues until some limit is reached...thus wasting resources and sometimes give a perceived okey end result. Other codes probably do the same). Having a `Monitor` sounds reasonable to me. But I also agree that designing a general user interface is not going to be very easy. But if we can do something like this:

- `Monitor` fetches the stdoutput or whatever the plugin defines during run, at configurable intervals (optionally we should also parse detailed scheduler info if the plugin wants)
- We define a monitor parser on the plugin level that parses this information and trigger some kind of error codes
- The plugin then needs to react

**Comment by yakutovicha** [comment]
> I take it the use-case for this would be to pre-emptively stop a molecular dynamics run

well, any long run that can potentially be restarted (MD, Geo Opt, NEB, ...). The issue here is that in a case of an abrupt termination a wavefunction file might be half-written, which won't allow restarting from it.

>  I think this is a feature that falls under a more general request...

Ok, good. Let's rename the issue than to generalise the request. Unless @espenfl has made another one already?

I leave the technical aspects aside because I am currently not in the position to comment on how this should be done, but I agree that this is useful. 

I have an additional proposition, though. For features like sending email when a job is completed (useful for AiiDA lab, for instance), it might be useful to register some methods that will be run after a process is completed.

Something like:
```
def send_emal():
     ...

p = Process()

p.on_completed(send_email)
```

I think this functionality also falls in the spirit of the request. Mentioning @giovannipizzi as we already had a preliminary discussion on that.

**Comment by sphuber** [comment]
The last part, regarding triggers on process termination are already possible, but it has to be implemented on the process class. For example you can already do:
```
class CP2KCalculation(CalcJob):

    def on_terminated(self):
        super().on_terminated()
        try:
            send_email(process_state=self.process_state)
        except Exception:
            self.logger.exception()
```
But this is of course not really useful if one has to enable this for all processes. We should probably then implement this on the base `Process` class and allow it to be enabled on a per process instance basis through the inputs. Maybe we can add it to the metadata:
```
spec.input('metadata.on_terminate.notification_send', default=False)
spec.input('metadata.on_terminate.notification_email', required=False, help='Email address to send notifications to. If not specified the email of the owner of the process will be used')
```

**Comment by yakutovicha** [comment]
There is another use case that falls in the topic's category. In AiiDA lab we provide a widget that shows the output of a process on the fly. Currently, we have a very [basic implementation](https://github.com/aiidalab/aiidalab-widgets-base/blob/e9b11c54e62a0f37cfa78314c6e7d1e77426abd6/aiidalab_widgets_base/process.py#L162) that gets the output of a process running on a local computer. @csadorf is working on adding the possibility to fetch output through ssh.

To make things more efficient it would make sense to allow do such requests of the output file directly through AiiDA core, as AiiDA has all the necessary machinery.

**Comment by giovannipizzi** [comment]
RemoteData has methods to get the list and content of files. Maybe one should try to use those?
Two notes:
- The AiiDA lab app should be responsible of opening the ssh connection and keeping it open (e.g. by opening the AiiDA transport) until the end, and not reopening every few seconds. This is a critical requirement to avoid to have the computer centre disable your account if they see too many connections
- Probably you might need a function to get the content of a file *from a given offset*, for a given length (or up to the end)? If the file is big, and you now it's being created and appended and not rewritten, you want to fetch only the new data at every request, otherwise you end up transferring too much data

Pinging @csadorf for info

**Comment by sphuber** [comment]
This feature request is now fulfilled by the [calcjob monitoring feature](https://aiida.readthedocs.io/projects/aiida-core/en/v2.4.0/howto/run_codes.html#how-to-monitor-and-prematurely-stop-a-calculation)
